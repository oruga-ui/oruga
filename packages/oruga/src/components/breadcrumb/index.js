import Breadcrumb from './Breadcrumb'
import BreadcrumbItem from './BreadcrumbItem'

import { use, registerComponent } from '../../utils/plugins'

const Plugin = {
    install(Vue) {
        registerComponent(Vue, Breadcrumb)
        registerComponent(Vue, BreadcrumbItem)
    }
}

use(Plugin)

export default Plugin

export {
    Breadcrumb as OBreadcrumb,
    BreadcrumbItem as OBreadcrumbItem
}
