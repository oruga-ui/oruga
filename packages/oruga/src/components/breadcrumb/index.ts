import type { App, Plugin } from "vue";

import Breadcrumb from "./Breadcrumb.vue";
import BreadcrumbItem from "./BreadcrumbItem.vue";

import { registerComponent } from "@/utils/plugins";

/** export breadcrumb plugin */
export default {
    install(app: App) {
        registerComponent(app, Breadcrumb);
        registerComponent(app, BreadcrumbItem);
    },
} as Plugin;

/** export Breadcrumb components */
export { Breadcrumb as OBreadcrumb, BreadcrumbItem as OBreadcrumbItem };
