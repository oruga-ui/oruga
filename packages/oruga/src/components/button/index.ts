import type { App, Plugin } from "vue";

import Button from "./Button.vue";

import { registerComponent } from "@/utils/plugins";

/** export button plugin */
export default {
    install(app: App) {
        registerComponent(app, Button);
    },
} as Plugin;

/** export button components */
export { Button as OButton };
