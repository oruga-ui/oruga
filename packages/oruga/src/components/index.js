import Autocomplete from './autocomplete'
import Button from './button'
import Checkbox from './checkbox'
import Collapse from './collapse'
import Datepicker from './datepicker'
import Dropdown from './dropdown'
import Field from './field'
import Icon from './icon'
import Input from './input'
import Loading from './loading'
import Modal from './modal'
import Pagination from './pagination'
import Radio from './radio'
import Select from './select'
import Skeleton from './skeleton'
import Sidebar from './sidebar'
import Slider from './slider'
import Steps from './steps'
import Switch from './switch'
import Table from './table'
import Tabs from './tabs'
import Tooltip from './tooltip'
import Upload from './upload'

export {
    Autocomplete,
    Button,
    Checkbox,
    Collapse,
    Datepicker,
    Dropdown,
    Field,
    Icon,
    Input,
    Loading,
    Modal,
    Pagination,
    Radio,
    Select,
    Skeleton,
    Sidebar,
    Slider,
    Steps,
    Switch,
    Table,
    Tabs,
    Tooltip,
    Upload
}
