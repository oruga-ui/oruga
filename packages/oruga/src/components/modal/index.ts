import type { App, Plugin } from "vue";

import Modal from "./Modal.vue";
import useModalProgrammatic from "./useModalProgrammatic";

import {
    registerComponent,
    registerComponentProgrammatic,
} from "@/utils/plugins";

/** export modal specific types */
// no types to export here

/** export modal plugin */
export default {
    install(app: App) {
        registerComponent(app, Modal);
        registerComponentProgrammatic(app, "modal", useModalProgrammatic);
    },
} as Plugin;

/** export modal components & composables */
export { Modal as OModal, useModalProgrammatic };
