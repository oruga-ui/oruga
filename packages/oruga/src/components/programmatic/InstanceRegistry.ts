import type { ComponentInternalInstance } from "vue";

export default class InstanceRegistry<T = ComponentInternalInstance> {
    entries: Array<T>;

    constructor() {
        this.entries = [];
    }

    /** Returns the number of registered active instances. */
    count(): number {
        return this.entries.length;
    }

    /** Returns the first registered active instance. */
    fist(): T | undefined {
        return this.entries.at(0);
    }

    /** Returns the last registered active instance. */
    last(): T | undefined {
        return this.entries.at(-1);
    }

    /** Adds a new instance to the instance stack. */
    add(entry: T): void {
        this.entries.push(entry);
    }

    /** Removes an instance from the instance stack. */
    remove(entry: T): void {
        const index = this.entries.indexOf(entry);
        this.entries.splice(index, 1);
    }

    /** Call a function for every registered active instance. */
    walk(callback: (value: T) => boolean | void): void {
        // Walk a copy of the array so that the callback is allowed to remove the instance
        this.entries = [...this.entries].filter((e) => {
            const ret = callback(e);
            return !(ret === true);
        });
    }
}
