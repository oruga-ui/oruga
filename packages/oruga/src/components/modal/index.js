import Modal from './Modal'

import { VueInstance } from '../../utils/config'
import { merge } from '../../utils/helpers'
import { use, registerComponent, registerComponentProgrammatic } from '../../utils/plugins'
import InstanceRegistry from '../../utils/InstanceRegistry'

let localVueInstance

let instances = new InstanceRegistry()

const ModalProgrammatic = {
    open(params) {
        let parent
        if (typeof params === 'string') {
            params = {
                content: params
            }
        }

        const defaultParam = {
            programmatic: { instances }
        }
        if (params.parent) {
            parent = params.parent
            delete params.parent
        }
        let slot
        if (Array.isArray(params.content)) {
            slot = params.content
            delete params.content
        }
        const propsData = merge(defaultParam, params)

        if (window.Promise) {
            propsData.promise = new Promise((p1, p2) => {
                propsData.programmatic.resolve = p1
                propsData.programmatic.reject = p2
            })
        }

        const vm = typeof window !== 'undefined' && window.Vue ? window.Vue : localVueInstance || VueInstance
        const ModalComponent = vm.extend(Modal)
        const instance = new ModalComponent({
            parent,
            el: document.createElement('div'),
            propsData
        })
        if (slot) {
            instance.$slots.default = slot
        }
        return instance
    },
    closeAll() {
        instances.walk((entry) => {
            entry.close(...arguments)
        })
    }
}

const Plugin = {
    install(Vue) {
        localVueInstance = Vue
        registerComponent(Vue, Modal)
        registerComponentProgrammatic(Vue, 'modal', ModalProgrammatic)
    }
}

use(Plugin)

export default Plugin

export {
    ModalProgrammatic,
    Modal as OModal
}
