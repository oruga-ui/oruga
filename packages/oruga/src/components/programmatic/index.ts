import type { App, Plugin } from "vue";

import Programmatic from "./useProgrammatic";

import { registerComponentProgrammatic } from "@/utils/plugins";
import InstanceRegistry from "./InstanceRegistry";

/** export programmatic specific types */
export type {
    ProgrammaticOptions,
    ProgrammaticExpose,
    PublicProgrammaticComponentOptions,
} from "./useProgrammatic";
export { InstanceRegistry };

/** export programmatic plugin */
export default {
    install(app: App) {
        registerComponentProgrammatic(app, "programmatic", Programmatic);
    },
} as Plugin;

/** export composables here */
export { useProgrammatic } from "./useProgrammatic";
