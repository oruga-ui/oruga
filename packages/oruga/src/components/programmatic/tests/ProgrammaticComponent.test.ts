import { createVNode, markRaw } from "vue";
import { describe, test, expect, afterEach, vi } from "vitest";
import { enableAutoUnmount, flushPromises, mount } from "@vue/test-utils";

import { OModal } from "@/components/modal";

import InstanceRegistry from "../InstanceRegistry";
import {
    ProgrammaticComponent,
    type ProgrammaticComponentExpose,
} from "../ProgrammaticComponent";

describe("ProgrammaticComponent tests", () => {
    enableAutoUnmount(afterEach);

    test("test render simple div correctly", () => {
        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: "div",
                props: { "data-oruga": "programmatic" },
            },
        });

        expect(!!wrapper.vm).toBeTruthy();
        expect(wrapper.exists()).toBeTruthy();
        expect(wrapper.html()).toMatchSnapshot();
        const element = wrapper.find('[data-oruga="programmatic"]');
        expect(element.exists()).toBeTruthy();
    });

    test("test render slot correctly", () => {
        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: createVNode({
                    template: `<div><slot /></div>`,
                }),
                props: { "data-oruga": "programmatic" },
            },
            slots: {
                default: `<p data-oruga="inner-slot">HELP</p>`,
            },
        });

        const element = wrapper.find('[data-oruga="programmatic"]');
        expect(element.exists()).toBeTruthy();
        const inner = wrapper.find('[data-oruga="inner-slot"]');
        expect(inner.exists()).toBeTruthy();
    });

    test("test render complex component with props correctly", () => {
        const content = "This is my content";
        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: markRaw(OModal),
                props: { content },
            },
        });

        const model = wrapper.find('[data-oruga="modal"]');
        expect(model.exists()).toBeTruthy();
        expect(model.text()).toBe(content);
    });

    test("test close is called correctly", async () => {
        vi.useFakeTimers();

        const onClose = vi.fn();
        const onDestroy = vi.fn();
        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: createVNode({
                    template: `<button @click="$emit('close', 'abc')"></button>`,
                }),
                props: { "data-oruga": "programmatic" },
                onClose,
                onDestroy,
            },
        });

        const button = wrapper.find("button");
        await button.trigger("click");

        vi.runAllTimers();

        const closeEmits = wrapper.emitted("close");
        expect(closeEmits).toHaveLength(1);
        expect(closeEmits[0][0]).toBe("abc");
        const destroyEmits = wrapper.emitted("destroy");
        expect(destroyEmits).toHaveLength(1);

        expect(onClose).toHaveBeenCalledOnce();
        expect(onDestroy).toHaveBeenCalledOnce();

        vi.useRealTimers();
    });

    test("test promise is called correctly", async () => {
        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: "div",
                props: { "data-oruga": "programmatic" },
            },
        });

        const component = wrapper.vm as unknown as ProgrammaticComponentExpose;
        expect(component.promise).not.toBeUndefined();

        // check promise get called
        const handler = vi.fn();
        component.promise.then(() => handler());
        expect(handler).not.toHaveBeenCalled();

        component.close(); // call close programmaticaly

        const closeEmits = wrapper.emitted("close");
        expect(closeEmits).toHaveLength(1);

        await flushPromises(); // await promise finished
        expect(handler).toHaveBeenCalledOnce();
    });

    test("test instance registry is called correctly", async () => {
        const instanceRegistry = new InstanceRegistry();

        expect(instanceRegistry.entries).toHaveLength(0);

        const wrapper = mount(ProgrammaticComponent, {
            props: {
                component: "div",
                instances: instanceRegistry,
            },
        });

        expect(instanceRegistry.entries).toHaveLength(1);

        wrapper.unmount();

        expect(instanceRegistry.entries).toHaveLength(0);
    });
});
