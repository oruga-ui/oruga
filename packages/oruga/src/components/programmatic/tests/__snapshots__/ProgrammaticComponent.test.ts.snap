// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProgrammaticComponent tests > test render simple div correctly 1`] = `"<div data-oruga="programmatic"></div>"`;
